### PR TITLE
jobs: add RevertFailed terminal state for jobs that fail during revert

### LINKED ISSUE #59542: jobs: add new terminal state indicating failure in `OnFailOrCancel`
**Describe the problem**

When jobs fail in `OnFailOrCancel()` almost always it's a sign that we did something wrong. It's bad that we then find ourselves looking at this issues with no history of what happened. We have no history because jobs get GC'd from terminal states. There is a hazard that if we don't GC these jobs that they'll fill up the jobs table, but honestly, that seems good because it will be a forcing function to uncover the fact that there are serious problems afoot. 

The proposal is to add some new terminal job state that indicates that the job failed to process its `OnFailOrCancel` hook and thus should stick around until somebody deals with it.

This is especially what I'd like to see for the cases of known corruption due to rollbacks of schema changes which drop columns that have constraints like outlined in https://github.com/cockroachdb/cockroach/issues/46541.  This is also potentially what we want for https://github.com/cockroachdb/cockroach/issues/59415. We should probably file a sentry report in these cases too.

### PR DESCRIPTION
**Pull request by adityamaru**
Previously, a failure during revert would mark the job as "failed" and
indicate in the error string that manual cleanup is required. This job
was however gc'eable if the retention time has elapsed.  This would
leave no trace of the job which is bad, since we expect users/support to
cleanup the partial work done by the job.

This change adds a new terminal state `RevertFailed` that does not allow
the job to get gc'ed.

Fixes: #59542

Release note: None

### CODE REVIEW

**Review by ajwerner — APPROVED** [review]
:lgtm: nice

> Reviewed 5 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63311#-:-MY80F8_A9MpVYWt7uyQ:bb76hxo)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @pbardea)

File: pkg/jobs/jobs.go [hunk 1]
```diff
@@ -172,7 +176,7 @@ func deprecatedIsOldSchemaChangeJob(payload *jobspb.Payload) bool {
 // Terminal returns whether this status represents a "terminal" state: a state
 // after which the job should never be updated again.
 func (s Status) Terminal() bool {
-	return s == StatusFailed || s == StatusSucceeded || s == StatusCanceled
+	return s == StatusFailed || s == StatusSucceeded || s == StatusCanceled || s == StatusRevertFailed
```
Comment (pbardea) [inline pkg/jobs/jobs.go:179]: Given this job state is treated as `Terminal`, should they also be considered a candidate for GC?

Comment (ajwerner) [reply pkg/jobs/jobs.go:179]: A major motivation here is to not GC these things.

Comment (pbardea) [reply pkg/jobs/jobs.go:179]: Ah, I had incorrectly thought that we check if a job is Terminal to determine if it's a candidate for GC, but that's not the case. Sorry for the noise.

**Comment by adityamaru** [comment]
TFTRs!

bors r=ajwerner,pbardea

**Comment by adityamaru** [comment]
Oops @pbardea I just realized you hadn't LGTM'ed this. If you have any comments I'll address them separately!

**Comment by adityamaru** [comment]
rebased on master

**Comment by adityamaru** [comment]
Dropped off my radar. Updated the test to not check for the orphaned jobs being cleaned up because we no longer have this logic after - https://github.com/cockroachdb/cockroach/pull/63730.

**Pull request merged**